src/utils/fusionAuth.ts

```diff
@@ -0,0 +1,165 @@
+import fs from 'fs-extra';
+import os from 'os';
+import path from 'path';
+import prompts from 'prompts';
+import Webhead, { WebheadInstance, WebheadRequestOptions } from 'webhead';
+
+const builderApiURL = '{HOST}/api/builder';
+const fusionAuthURL = 'https://fusionauth-ce.betty.services';
+
+type LoginResponse = {
+  token: string;
+  refreshToken: string;
+  twoFactorId: string;
+};
+
+type TwoFactorLoginResponse = {
+  token: string;
+  refreshToken: string;
+};
+
+type UserResponse = {
+  user: object;
+};
+
+type Response = LoginResponse | TwoFactorLoginResponse | UserResponse;
+
+class FusionAuth {
+  private configFile: string;
+
+  private host: string;
+
+  public loginId?: string;
+
+  public password?: string;
+
+  private relogin: () => Promise<void>;
+
+  private webhead: WebheadInstance;
+
+  constructor(host: string, relogin: () => Promise<void>) {
+    this.configFile = path.join(os.homedir(), '.bb-cli-fa');
+    this.host = host;
+    this.relogin = relogin;
+    this.webhead = Webhead();
+  }
+
+  async ensureLogin(): Promise<void> {
+    const response = (await this.get('/api/user', {
+      headers: {
+        Authorization: `Bearer ${this.jwt()}`,
+      },
+    })) as UserResponse;
+
+    if (!response || !response.user) {
+      if (!this.loginId) {
+        await this.relogin();
+      }
+      await this.login();
+    }
+  }
+
+  async login(): Promise<void> {
+    const { token, refreshToken, twoFactorId } = (await this.post(
+      '/api/login',
+      {
+        json: {
+          loginId: this.loginId,
+          password: this.password,
+        },
+      },
+    )) as LoginResponse;
+
+    if (token) {
+      this.storeTokens(token, refreshToken);
+    } else if (twoFactorId) {
+      await this.complete2FA(twoFactorId);
+    }
+  }
+
+  async complete2FA(twoFactorId: string): Promise<void> {
+    const { code } = await prompts([
+      {
+        type: 'text',
+        name: 'code',
+        message: 'Fill in your 2FA code (to upload code)',
+      },
+    ]);
+
+    const { token, refreshToken } = (await this.post('/api/two-factor/login', {
+      json: {
+        code,
+        twoFactorId,
+      },
+    })) as TwoFactorLoginResponse;
+
+    if (token) {
+      this.storeTokens(token, refreshToken);
+    } else {
+      await this.complete2FA(twoFactorId);
+    }
+  }
+
+  async upload(uuid: string, zipFile: string): Promise<boolean> {
+    await this.ensureLogin();
+
+    const { statusCode } = await this.webhead.post(
+      `${this.builderApiURL()}/artifacts/actions/${uuid}/functions`,
+      {
+        headers: {
+          Authorization: `Bearer ${this.jwt()}`,
+        },
+        multiPartData: [{ name: 'file', file: zipFile }],
+      },
+    );
+
+    return !!statusCode.toString().match(/^2/);
+  }
+
+  async get(
+    urlPath: string,
+    options: WebheadRequestOptions,
+  ): Promise<Response> {
```

Comment:
Instead of using `Response` as a union here, I think it'll be better to use generics and then pass the object you expect to retrieve from the function.

E.g.

```javascript
async get<T>(
    urlPath: string,
    options: WebheadRequestOptions,
  ): Promise<T> {
    ...
```

Now you can call it like this:

```javascript
const response =  await this.get<UserResponse>('/api/user', {
      headers: {
        Authorization: `Bearer ${this.jwt()}`,
      },
    })
```

TypeScript will now know that the type of `response` is `UserResponse`. No need for the `as ...` anymore.

Reply:
Same goes for the `post` and `request` functions.